File: src/test/java/br/com/useinet/finance/controller/TransacaoControllerIT.java
package br.com.useinet.finance.controller;

import br.com.useinet.finance.dto.TransacaoResponse;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseToken;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.jdbc.core.JdbcTemplate;

import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
class TransacaoControllerIT {

    @Autowired
    private TestRestTemplate restTemplate;

    @Autowired
    private JdbcTemplate jdbcTemplate;

    @MockBean
    private FirebaseAuth firebaseAuth;

    static final String MOCK_TOKEN = "mock-firebase-token";

    @BeforeEach
    void setUp() throws Exception {
        cleanupUser("carlos.transacao@it.com");

        FirebaseToken mockToken = mock(FirebaseToken.class);
        when(mockToken.getUid()).thenReturn("uid-transacao");
        when(mockToken.getEmail()).thenReturn("carlos.transacao@it.com");
        when(mockToken.getName()).thenReturn("Carlos");
        when(firebaseAuth.verifyIdToken(eq(MOCK_TOKEN))).thenReturn(mockToken);
    }

    private void cleanupUser(String email) {
        jdbcTemplate.update("DELETE FROM transacoes WHERE usuario_id IN (SELECT id FROM usuarios WHERE email = ?)", email);
        jdbcTemplate.update("DELETE FROM contas WHERE usuario_id IN (SELECT id FROM usuarios WHERE email = ?)", email);
        jdbcTemplate.update("DELETE FROM usuarios WHERE email = ?", email);
    }

    private HttpHeaders authHeaders() {
        HttpHeaders headers = new HttpHeaders();
        headers.setBearerAuth(MOCK_TOKEN);
        return headers;
    }

    @Test
    void criar_shouldReturn201AndTransacao() {
        var request = Map.of(
                "descricao", "Salário",
                "valor", 5000.0,
                "tipo", "RECEITA"
        );

        ResponseEntity<TransacaoResponse> response = restTemplate.exchange(
                "/transactions",
                HttpMethod.POST,
                new HttpEntity<>(request, authHeaders()),
                TransacaoResponse.class
        );

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.CREATED);
        assertThat(response.getBody().getDescricao()).isEqualTo("Salário");
        assertThat(response.getBody().getValor()).isEqualTo(5000.0);
    }

    @Test
    void listar_shouldReturnTransacoesDoUsuario() {
        var request = Map.of("descricao", "Mercado", "valor", 200.0, "tipo", "DESPESA");
        restTemplate.exchange("/transactions", HttpMethod.POST, new HttpEntity<>(request, authHeaders()), TransacaoResponse.class);

        ResponseEntity<TransacaoResponse[]> response = restTemplate.exchange(
                "/transactions",
                HttpMethod.GET,
                new HttpEntity<>(authHeaders()),
                TransacaoResponse[].class
        );

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(response.getBody()).isNotEmpty();
    }

    @Test
    void deletar_shouldReturn204() {
        var request = Map.of("descricao", "A deletar", "valor", 50.0, "tipo", "DESPESA");
        ResponseEntity<TransacaoResponse> created = restTemplate.exchange(
                "/transactions", HttpMethod.POST, new HttpEntity<>(request, authHeaders()), TransacaoResponse.class
        );

        Long id = created.getBody().getId();

        ResponseEntity<Void> response = restTemplate.exchange(
                "/transactions/" + id,
                HttpMethod.DELETE,
                new HttpEntity<>(authHeaders()),
                Void.class
        );

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.NO_CONTENT);
    }

    @Test
    void listar_shouldReturn401WithoutToken() {
        ResponseEntity<String> response = restTemplate.getForEntity("/transactions", String.class);
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.UNAUTHORIZED);
    }

    @Test
    void atualizar_shouldReturn200WithUpdatedData() {
        var createReq = Map.of("descricao", "Original", "valor", 100.0, "tipo", "RECEITA");
        ResponseEntity<TransacaoResponse> created = restTemplate.exchange(
                "/transactions", HttpMethod.POST,
                new HttpEntity<>(createReq, authHeaders()), TransacaoResponse.class);

        Long id = created.getBody().getId();

        var updateReq = Map.of("descricao", "Atualizado", "valor", 200.0, "tipo", "RECEITA");
        ResponseEntity<TransacaoResponse> response = restTemplate.exchange(
                "/transactions/" + id, HttpMethod.PUT,
                new HttpEntity<>(updateReq, authHeaders()), TransacaoResponse.class);

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(response.getBody().getDescricao()).isEqualTo("Atualizado");
        assertThat(response.getBody().getValor()).isEqualTo(200.0);
    }

    @Test
    void listar_comFiltroData_shouldReturn200() {
        var createReq = Map.of("descricao", "Filtrado", "valor", 50.0, "tipo", "DESPESA");
        restTemplate.exchange("/transactions", HttpMethod.POST,
                new HttpEntity<>(createReq, authHeaders()), TransacaoResponse.class);

        ResponseEntity<TransacaoResponse[]> response = restTemplate.exchange(
                "/transactions?inicio=2020-01-01&fim=2099-12-31",
                HttpMethod.GET, new HttpEntity<>(authHeaders()), TransacaoResponse[].class);

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(response.getBody()).isNotNull();
    }

    @Test
    void listar_comDataInvalida_shouldReturn400() {
        ResponseEntity<String> response = restTemplate.exchange(
                "/transactions?inicio=2025-06-01&fim=2025-01-01",
                HttpMethod.GET, new HttpEntity<>(authHeaders()), String.class);

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.BAD_REQUEST);
    }

    @Test
    void exportarCsv_shouldReturn200WithCsvContentType() {
        var createReq = Map.of("descricao", "CSV Export", "valor", 300.0, "tipo", "RECEITA");
        restTemplate.exchange("/transactions", HttpMethod.POST,
                new HttpEntity<>(createReq, authHeaders()), TransacaoResponse.class);

        ResponseEntity<byte[]> response = restTemplate.exchange(
                "/transactions/export/csv", HttpMethod.GET,
                new HttpEntity<>(authHeaders()), byte[].class);

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(response.getBody()).isNotNull();
        assertThat(response.getHeaders().getContentType().toString()).contains("text/csv");
    }

    @Test
    void exportarCsv_shouldContainContaColumn() {
        var createReq = Map.of("descricao", "CSV Conta", "valor", 100.0, "tipo", "DESPESA");
        restTemplate.exchange("/transactions", HttpMethod.POST,
                new HttpEntity<>(createReq, authHeaders()), TransacaoResponse.class);

        ResponseEntity<byte[]> response = restTemplate.exchange(
                "/transactions/export/csv", HttpMethod.GET,
                new HttpEntity<>(authHeaders()), byte[].class);

        String csv = new String(response.getBody(), java.nio.charset.StandardCharsets.UTF_8);
        assertThat(csv).contains("Conta");
    }

    @Test
    void criar_comConta_shouldReturnContaIdNaResposta() {
        var contaReq = Map.of("nome", "Nubank", "saldo", 1000.0);
        ResponseEntity<Map> contaResp = restTemplate.exchange(
                "/contas", HttpMethod.POST,
                new HttpEntity<>(contaReq, authHeaders()), Map.class);
        Integer contaId = (Integer) contaResp.getBody().get("id");

        var request = Map.of("descricao", "Mercado", "valor", 200.0, "tipo", "DESPESA", "contaId", contaId);
        ResponseEntity<TransacaoResponse> response = restTemplate.exchange(
                "/transactions", HttpMethod.POST,
                new HttpEntity<>(request, authHeaders()), TransacaoResponse.class);

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.CREATED);
        assertThat(response.getBody().getContaId()).isEqualTo(contaId.longValue());
        assertThat(response.getBody().getConta()).isEqualTo("Nubank");
    }
}

File: src/test/kotlin/br/com/useinet/finance/controller/TransacaoControllerIT.kt
package br.com.useinet.finance.controller

import br.com.useinet.finance.dto.TransacaoResponse
import br.com.useinet.finance.support.IntegrationTestBase
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import com.google.firebase.auth.FirebaseToken
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito.eq
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when`
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.http.*
import java.nio.charset.StandardCharsets

class TransacaoControllerIT : IntegrationTestBase() {

    @Autowired lateinit var restTemplate: TestRestTemplate

    companion object {
        const val MOCK_TOKEN = "mock-firebase-token"
    }

    @BeforeEach
    fun setUp() {
        cleanupUser("carlos.transacao@it.com")
        val mockToken = mock(FirebaseToken::class.java)
        `when`(mockToken.uid).thenReturn("uid-transacao")
        `when`(mockToken.email).thenReturn("carlos.transacao@it.com")
        `when`(mockToken.name).thenReturn("Carlos")
        `when`(firebaseAuth.verifyIdToken(eq(MOCK_TOKEN))).thenReturn(mockToken)
    }

    private fun authHeaders() = HttpHeaders().apply { setBearerAuth(MOCK_TOKEN) }

    @Test
    fun criar_shouldReturn201AndTransacao() {
        val response = restTemplate.exchange(
            "/transactions", HttpMethod.POST,
            HttpEntity(mapOf("descricao" to "Salário", "valor" to 5000.0, "tipo" to "RECEITA"), authHeaders()),
            TransacaoResponse::class.java
        )
        assertThat(response.statusCode).isEqualTo(HttpStatus.CREATED)
        assertThat(response.body!!.descricao).isEqualTo("Salário")
        assertThat(response.body!!.valor).isEqualTo(5000.0)
    }

    @Test
    fun listar_shouldReturnTransacoesDoUsuario() {
        restTemplate.exchange("/transactions", HttpMethod.POST, HttpEntity(mapOf("descricao" to "Mercado", "valor" to 200.0, "tipo" to "DESPESA"), authHeaders()), TransacaoResponse::class.java)
        val response = restTemplate.exchange("/transactions", HttpMethod.GET, HttpEntity<Any>(authHeaders()), String::class.java)
        assertThat(response.statusCode).isEqualTo(HttpStatus.OK)
        val page = jacksonObjectMapper().readValue<Map<String, Any>>(response.body!!)
        assertThat(page["content"] as List<*>).isNotEmpty
    }

    @Test
    fun deletar_shouldReturn204() {
        val created = restTemplate.exchange(
            "/transactions", HttpMethod.POST,
            HttpEntity(mapOf("descricao" to "A deletar", "valor" to 50.0, "tipo" to "DESPESA"), authHeaders()),
            TransacaoResponse::class.java
        )
        val id = created.body!!.id
        val response = restTemplate.exchange("/transactions/$id", HttpMethod.DELETE, HttpEntity<Any>(authHeaders()), Void::class.java)
        assertThat(response.statusCode).isEqualTo(HttpStatus.NO_CONTENT)
    }

    @Test
    fun listar_shouldReturn401WithoutToken() {
        val response = restTemplate.getForEntity("/transactions", String::class.java)
        assertThat(response.statusCode).isEqualTo(HttpStatus.UNAUTHORIZED)
    }

    @Test
    fun atualizar_shouldReturn200WithUpdatedData() {
        val created = restTemplate.exchange(
            "/transactions", HttpMethod.POST,
            HttpEntity(mapOf("descricao" to "Original", "valor" to 100.0, "tipo" to "RECEITA"), authHeaders()),
            TransacaoResponse::class.java
        )
        val id = created.body!!.id
        val response = restTemplate.exchange(
            "/transactions/$id", HttpMethod.PUT,
            HttpEntity(mapOf("descricao" to "Atualizado", "valor" to 200.0, "tipo" to "RECEITA"), authHeaders()),
            TransacaoResponse::class.java
        )
        assertThat(response.statusCode).isEqualTo(HttpStatus.OK)
        assertThat(response.body!!.descricao).isEqualTo("Atualizado")
        assertThat(response.body!!.valor).isEqualTo(200.0)
    }

    @Test
    fun listar_comFiltroData_shouldReturn200() {
        restTemplate.exchange("/transactions", HttpMethod.POST, HttpEntity(mapOf("descricao" to "Filtrado", "valor" to 50.0, "tipo" to "DESPESA"), authHeaders()), TransacaoResponse::class.java)
        val response = restTemplate.exchange("/transactions?inicio=2020-01-01&fim=2099-12-31", HttpMethod.GET, HttpEntity<Any>(authHeaders()), String::class.java)
        assertThat(response.statusCode).isEqualTo(HttpStatus.OK)
        val page = jacksonObjectMapper().readValue<Map<String, Any>>(response.body!!)
        // content must be a list and must include the transaction just created
        val content = page["content"] as List<*>
        assertThat(content).isNotEmpty
        val body = jacksonObjectMapper().writeValueAsString(content)
        assertThat(body).contains("Filtrado")
    }

    @Test
    fun listar_comDataInvalida_shouldReturn400() {
        val response = restTemplate.exchange("/transactions?inicio=2025-06-01&fim=2025-01-01", HttpMethod.GET, HttpEntity<Any>(authHeaders()), String::class.java)
        assertThat(response.statusCode).isEqualTo(HttpStatus.BAD_REQUEST)
    }

    @Test
    fun exportarCsv_shouldReturn200WithCsvContentType() {
        restTemplate.exchange("/transactions", HttpMethod.POST, HttpEntity(mapOf("descricao" to "CSV Export", "valor" to 300.0, "tipo" to "RECEITA"), authHeaders()), TransacaoResponse::class.java)
        val response = restTemplate.exchange("/transactions/export/csv", HttpMethod.GET, HttpEntity<Any>(authHeaders()), ByteArray::class.java)
        assertThat(response.statusCode).isEqualTo(HttpStatus.OK)
        assertThat(response.body).isNotNull
        assertThat(response.headers.contentType.toString()).contains("text/csv")
    }

    @Test
    fun exportarCsv_shouldContainContaColumn() {
        restTemplate.exchange("/transactions", HttpMethod.POST, HttpEntity(mapOf("descricao" to "CSV Conta", "valor" to 100.0, "tipo" to "DESPESA"), authHeaders()), TransacaoResponse::class.java)
        val response = restTemplate.exchange("/transactions/export/csv", HttpMethod.GET, HttpEntity<Any>(authHeaders()), ByteArray::class.java)
        val csv = String(response.body!!, StandardCharsets.UTF_8)
        // verify the full column header structure, not just the presence of "Conta"
        assertThat(csv).contains("ID;Descrição;Valor;Tipo;Data;Categoria;Conta;\"Saldo da Conta\"")
        assertThat(csv).contains("CSV Conta")
    }

    @Test
    fun criar_comConta_shouldReturnContaIdNaResposta() {
        val contaResp = restTemplate.exchange("/contas", HttpMethod.POST, HttpEntity(mapOf("nome" to "Nubank", "saldo" to 1000.0), authHeaders()), Map::class.java)
        val contaId = (contaResp.body!![("id")] as Number).toInt()
        val response = restTemplate.exchange(
            "/transactions", HttpMethod.POST,
            HttpEntity(mapOf("descricao" to "Mercado", "valor" to 200.0, "tipo" to "DESPESA", "contaId" to contaId), authHeaders()),
            TransacaoResponse::class.java
        )
        assertThat(response.statusCode).isEqualTo(HttpStatus.CREATED)
        assertThat(response.body!!.contaId).isEqualTo(contaId.toLong())
        assertThat(response.body!!.conta).isEqualTo("Nubank")
    }

    @Test
    fun listar_paginacao_shouldRespectPageSizeParam() {
        repeat(5) { i ->
            restTemplate.exchange("/transactions", HttpMethod.POST,
                HttpEntity(mapOf("descricao" to "Pag $i", "valor" to 10.0, "tipo" to "DESPESA"), authHeaders()),
                TransacaoResponse::class.java)
        }

        val response = restTemplate.exchange("/transactions?page=0&size=2", HttpMethod.GET, HttpEntity<Any>(authHeaders()), String::class.java)
        assertThat(response.statusCode).isEqualTo(HttpStatus.OK)
        val page = jacksonObjectMapper().readValue<Map<String, Any>>(response.body!!)
        val content = page["content"] as List<*>
        assertThat(content).hasSize(2)
        assertThat((page["totalElements"] as Number).toInt()).isGreaterThanOrEqualTo(5)
        assertThat((page["totalPages"] as Number).toInt()).isGreaterThanOrEqualTo(3)
    }

    @Test
    fun exportarCsv_shouldShowRunningBalancePerTransaction() {
        // Conta com saldo=0 para não gerar transação de saldo inicial
        val contaResp = restTemplate.postForEntity(
            "/contas",
            HttpEntity(mapOf("nome" to "ExtratoIT", "saldo" to 0.0), authHeaders()),
            Map::class.java
        )
        val contaId = (contaResp.body!!["id"] as Number).toInt()

        restTemplate.exchange("/transactions", HttpMethod.POST,
            HttpEntity(mapOf("descricao" to "Entrada", "valor" to 1000.0, "tipo" to "RECEITA", "contaId" to contaId), authHeaders()),
            TransacaoResponse::class.java)
        restTemplate.exchange("/transactions", HttpMethod.POST,
            HttpEntity(mapOf("descricao" to "Saida", "valor" to 300.0, "tipo" to "DESPESA", "contaId" to contaId), authHeaders()),
            TransacaoResponse::class.java)

        val csv = String(
            restTemplate.exchange("/transactions/export/csv", HttpMethod.GET, HttpEntity<Any>(authHeaders()), ByteArray::class.java).body!!,
            StandardCharsets.UTF_8
        )

        // Saldo acumulado após RECEITA 1000 = 1000, após DESPESA 300 = 700
        assertThat(csv).contains("ExtratoIT;1000,00")
        assertThat(csv).contains("ExtratoIT;700,00")
        assertThat(csv).doesNotContain("ExtratoIT;1300,00") // saldo atual incorreto
    }

    @Test
    fun listar_shouldNotReturnOtherUsersTransacoes() {
        val otherToken = "mock-other-transacao-token"
        val otherFirebaseToken = mock(FirebaseToken::class.java)
        `when`(otherFirebaseToken.uid).thenReturn("uid-other-transacao")
        `when`(otherFirebaseToken.email).thenReturn("other.transacao.isolation@it.com")
        `when`(otherFirebaseToken.name).thenReturn("Other Transacao User")
        `when`(firebaseAuth.verifyIdToken(eq(otherToken))).thenReturn(otherFirebaseToken)
        val otherHeaders = HttpHeaders().apply { setBearerAuth(otherToken) }

        restTemplate.exchange("/transactions", HttpMethod.POST,
            HttpEntity(mapOf("descricao" to "Other Secret Transaction", "valor" to 9999.0, "tipo" to "RECEITA"), otherHeaders),
            Any::class.java)

        val response = restTemplate.exchange("/transactions", HttpMethod.GET, HttpEntity<Any>(authHeaders()), String::class.java)
        assertThat(response.body).doesNotContain("Other Secret Transaction")
    }
}
